From 3d3476456f0fe1daef7f896e8351ddb2258f4353 Mon Sep 17 00:00:00 2001
From: Ricardo Salveti <ricardo@foundries.io>
Date: Fri, 22 Aug 2025 16:43:39 -0300
Subject: [PATCH] pkcs11: make code compatible with libp11 0.4.16

Update the pkcs11 keygen logic to reflect the changes done as part of
the upstream libp11 0.4.16 release.

Upstream-Status: Submitted [https://github.com/foundriesio/lmp-device-register/pull/62]

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>
---
 inc/device_register.h |  1 +
 src/pkcs11.cpp        | 12 +++++++++---
 2 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/inc/device_register.h b/inc/device_register.h
index a5ab30d..6c37ac3 100644
--- a/inc/device_register.h
+++ b/inc/device_register.h
@@ -27,6 +27,7 @@
 #include <openssl/err.h>
 #include <openssl/buffer.h>
 
+#include <boost/algorithm/hex.hpp>
 #include <boost/algorithm/string.hpp>
 #include <boost/beast/core/detail/base64.hpp>
 #include <boost/filesystem.hpp>
diff --git a/src/pkcs11.cpp b/src/pkcs11.cpp
index 340e4da..80176d7 100644
--- a/src/pkcs11.cpp
+++ b/src/pkcs11.cpp
@@ -239,6 +239,8 @@ int pkcs11_create_csr(const lmp_options &opt, string &pkey, string &csr)
 	unsigned int n = 0;
 	int ret = 0;
 	bool created_keys = false;
+	std::vector<unsigned char> tls_id_bytes;
+	PKCS11_params params = {.extractable = 0, .sensitive = 1};
 	PKCS11_EC_KGEN ec = {
 		.curve = "P-256",
 	};
@@ -246,14 +248,18 @@ int pkcs11_create_csr(const lmp_options &opt, string &pkey, string &csr)
 		.type = EVP_PKEY_EC,
 		.token_label = hsm_cfg.token.c_str(),
 		.key_label = hsm_cfg.tls_lbl.c_str(),
-		.key_id = hsm_cfg.tls_id.c_str(),
 	};
 	bool init = false;
 
+	boost::algorithm::unhex(hsm_cfg.tls_id, std::back_inserter(tls_id_bytes));
+	attr.key_id = tls_id_bytes.data();
+	attr.id_len = tls_id_bytes.size();
+
 	if (opt.hsm_module.empty())
 		leave;
 
 	attr.kgen.ec = &ec;
+	attr.key_params = &params;
 again:
 	ctx = PKCS11_CTX_new();
 
@@ -293,7 +299,7 @@ again:
 	if (PKCS11_login(slot, 0, opt.hsm_pin.c_str()))
 		leave_exit;
 
-	if (PKCS11_generate_key(slot->token, &attr))
+	if (PKCS11_keygen(slot->token, &attr))
 		leave_exit;
 
 	/* Keys have been created */
@@ -537,4 +543,4 @@ int pkcs11_cleanup(lmp_options &opt)
 	PKCS11_CTX_free(ctx);
 
 	return 0;
-}
\ No newline at end of file
+}
-- 
2.34.1

